nav2_route/src/plugins/graph_file_loaders/geojson_graph_file_loader.cpp

```diff
@@ -0,0 +1,242 @@
+  std::vector<Json> nodes, edges;
+  getGraphElements(features, nodes, edges);
+
+  if (nodes.empty() || edges.empty()) {
+    RCLCPP_ERROR(
+      logger_, "The graph is malformed. Is does not contain nodes or edges. Please check %s",
+      filepath.c_str());
+    return false;
+  }
+
+  graph.resize(nodes.size());
+  addNodesToGraph(graph, graph_to_id_map, nodes);
+  addEdgesToGraph(graph, graph_to_id_map, edges);
+  return true;
+}
+
+bool GeoJsonGraphFileLoader::doesFileExist(const std::string & filepath)
+{
+  return std::filesystem::exists(filepath);
+}
+
+
+void GeoJsonGraphFileLoader::getGraphElements(
+  const Json & features, std::vector<Json> & nodes, std::vector<Json> & edges)
+{
+  for (const auto & feature : features) {
+    if (feature["geometry"]["type"] == "Point") {
+      nodes.emplace_back(feature);
+    } else if (feature["geometry"]["type"] == "MultiLineString") {
+      edges.emplace_back(feature);
+    }
+  }
+}
+
+void GeoJsonGraphFileLoader::addNodesToGraph(
+  Graph & graph, GraphToIDMap & graph_to_id_map, std::vector<Json> & nodes)
+{
+  int idx = 0;
+  for (const auto & node : nodes) {
+    const auto properties = node["properties"];
+    graph[idx].nodeid = properties["id"];
+    graph_to_id_map[graph[idx].nodeid] = idx;
+    graph[idx].coords = convertCoordinatesFromJson(node);
+    graph[idx].operations = convertOperationsFromJson(properties);
+    graph[idx].metadata = convertMetaDataFromJson(properties);
+    idx++;
+  }
+}
+
+void GeoJsonGraphFileLoader::addEdgesToGraph(
+  Graph & graph, GraphToIDMap & graph_to_id_map, std::vector<Json> & edges)
+{
+  for (const auto & edge : edges) {
+    // Required data
+    const auto properties = edge["properties"];
+    unsigned int id = properties["id"];
+    unsigned int start_id = properties["startid"];
+    unsigned int end_id = properties["endid"];
+
+    if (graph_to_id_map.find(start_id) == graph_to_id_map.end()) {
+      RCLCPP_ERROR(logger_, "Start id %u does not exist for edge id %u", start_id, id);
+      throw nav2_core::NoValidGraph("Start id does not exist");
+    }
+
+    if (graph_to_id_map.find(end_id) == graph_to_id_map.end()) {
+      RCLCPP_ERROR(logger_, "End id of %u does not exist for edge id %u", end_id, id);
+      throw nav2_core::NoValidGraph("End id does not exist");
+    }
+
+    EdgeCost edge_cost = convertEdgeCostFromJson(properties);
+    Operations operations = convertOperationsFromJson(properties);
+    Metadata metadata = convertMetaDataFromJson(properties);
+
+    graph[graph_to_id_map[start_id]].addEdge(
+      edge_cost, &graph[graph_to_id_map[end_id]], id,
+      metadata, operations);
+  }
+}
+
+Coordinates GeoJsonGraphFileLoader::convertCoordinatesFromJson(const Json & node)
+{
+  Coordinates coords;
+  const auto & properties = node["properties"];
+  if (properties.contains("frame")) {
+    coords.frame_id = properties["frame"];
+  }
+
+  const auto & coordinates = node["geometry"]["coordinates"];
```

Comment:
Will the json throw or crash if there was, for instance, only a single point set in this vector and we try to access `[1]`?

Reply:
Yep.